pkg/ring/ring.go

```diff
@@ -34,16 +34,20 @@ const (
 
 	// CompactorRingKey is the key under which we store the compactors ring in the KVStore.
 	CompactorRingKey = "compactor"
+
+	// GetBufferSize is the suggested size of buffers passed to Ring.Get(). It's based on
+	// a typical replication factor 3, plus extra room for a JOINING + LEAVING instance.
+	GetBufferSize = 5
 )
 
 // ReadRing represents the read interface to the ring.
 type ReadRing interface {
 	prometheus.Collector
 
 	// Get returns n (or more) ingesters which form the replicas for the given key.
-	// buf is a slice to be overwritten for the return value
-	// to avoid memory allocation; can be nil.
-	Get(key uint32, op Operation, buf []IngesterDesc) (ReplicationSet, error)
+	// bufDescs, bufHosts and bufZones are slices to be overwritten for the return value
+	// to avoid memory allocation; can be nil, or created with ring.MakeBuffersForGet().
+	Get(key uint32, op Operation, bufDescs []IngesterDesc, bufHosts, bufZones []string) (ReplicationSet, error)
```

Comment:
Nit: it makes sense for caller to pass buffer for ingesters, because `Get` returns it back. Hosts/zones buffers are only optimizations, and using a ring-specific pool (shared with subrings) may be a good alternative, without a need to change `Get` call.

Reply:
I tried it, but unfortunately it makes is way slower on the hot path:

```
BenchmarkRing_Get-12                        853           1126          +32.00%
```

Reply:
Fair enough.